exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java

```diff
@@ -321,6 +329,38 @@ public void close() {
     return listener.getResults();
   }
 
+  public DrillRpcFuture<QueryPlanFragments> planQuery(QueryType type, String query, boolean isSplitPlan) {
+    GetQueryPlanFragments runQuery = GetQueryPlanFragments.newBuilder().setQuery(query).setType(type).setSplitPlan(isSplitPlan).build();
+    return client.submitPlanQuery(runQuery);
+  }
+
+  public void runQuery(QueryType type, List<PlanFragment> planFragments, UserResultsListener resultsListener)
+      throws RpcException {
+    // QueryType can be only executional
+    checkArgument((QueryType.EXECUTION == type), "Only EXECUTIONAL type query is supported with PlanFragments");
+    // setting Plan on RunQuery will be used for logging purposes and therefore can not be null
+    // since there is no Plan string provided we will create a JsonArray out of individual fragment Plans
+    ArrayNode jsonArray = objectMapper.createArrayNode();
+    for (PlanFragment fragment : planFragments) {
+      try {
+        jsonArray.add(objectMapper.readTree(fragment.getFragmentJson()));
+      } catch (IOException e) {
+        logger.error("Exception while trying to read PlanFragment JSON for %s", fragment.getHandle().getQueryId(), e);
+        throw new RpcException(e);
+      }
+    }
+    final String fragmentsToJsonString;
+    try {
+      fragmentsToJsonString = objectMapper.writeValueAsString(jsonArray);
```

Comment:
That's because we are not telling Jackson how to handle cyclicity. Can you try using DrillbitContext#getPlanReader instead?